### ISSUE TITLE
"no-cors" CSS SOP violation

### ISSUE THREAD

**Issue by annevk** [slightlyoff/ServiceWorker#719]
Per our current set of definitions a service worker reveals what resources a "no-cors" CSS stylesheet attached to a document loads. In particular this can leak confidential tokens in the URLs.

Entered the public record here: http://krijnhoetmer.nl/irc-logs/whatwg/20150703#l-286

According to @jakearchibald resource timing (paging @igrigorik) did this first, in both Chrome and Firefox.

I think we should revert both, seems like bad precedent to cut more holes in SOP.

**Comment by jakearchibald** [comment]
Most requests by CSS are exposed through computed styles, and it's pretty trivial to iterate over all elements to find those. Things like font urls & `@import` cannot be detected through computed styles, but are exposed by SW & resource timing.

I see CSS more like script, although it's "opaque", it gives up some visibility when it makes requests within the context of the page. Although, as @annevk points out, script gives up visibility by using globals that can be modified, whereas CSS doesn't.

**Comment by annevk** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1180145 tracks this for Gecko.

**Comment by jakearchibald** [comment]
@davidben @mikewest what's your take here, are we (and resource priorities) breaking the web by exposing CSS-initiated fetches?

**Comment by igrigorik** [comment]
Opened a bug on Resource Timing to track this (https://github.com/w3c/resource-timing/issues/27).

**Comment by jakearchibald** [comment]
Ping @davidben @mikewest

We have two options here:

1. Declare CSS-triggered fetches as "page visible" as they are with script-triggered fetches. However, this is leaking new data when it comes to URLs that don't appear in computed styles. Eg font sources, @import urls.
2. [as Anne suggests](https://bugzilla.mozilla.org/show_bug.cgi?id=1180145#c3), remove these from the resource timing API & bypass serviceworker for these requests.

This *will* trip developers up who host their CSS on a CDN, as it's less likely to have access-control-allow-origin, but it's not worth breaking the web over.

**Comment by igrigorik** [comment]
@jakearchibald silly question: if you hide CSS fetches from SW, how am I supposed to offline my app with background images and fonts? Wouldn't that effectively mean that "if you want to be SW-compatible, you should avoid CSS fetches"... because that seems like it would break the web. I don't see why we're making this distinction between JS and CSS (JS fetches are "computed fetches" also).

**Comment by annevk** [comment]
No-cors cross-origin CSS subresources would be hidden. Any other kind of CSS subresources would not be hidden. So there's still plenty of ways to do business.

**Comment by jakearchibald** [comment]
@igrigorik 

> if you hide CSS fetches from SW, how am I supposed to offline my app with background images and fonts?

CSS fetches would only be hidden from SW if the CSS is no-cors cross-origin. You'd be able to get them back with `<link rel="stylesheet" href="…" crossorigin>`, provided CORS headers.

**Comment by igrigorik** [comment]
CSS has no author-level equivalent for "crossorigin" for fetches declared via `url()` - i.e. I can't specify a CORS policy on resources specified _within_ the CSS file. As such, all `url()` initiated fetches (read, _almost all_ CSS fetches modulo some odd exceptions like web fonts), would become invisible? Unless I'm missing something here, that would break the web.

Why can SW see and intercept no-cors crossorigin `<img>` initiated fetches, but same use case then fails under CSS? This seems inconsistent and broken.

_p.s. the whole `crossorigin` business is a total mess - see https://github.com/w3c/resource-hints/issues/32._

**Comment by annevk** [comment]
No, it depends on how the CSS resource was fetched. Not how its subresources are fetched.

**Comment by igrigorik** [comment]
@annevk ah, ok.. So, coming back to @jakearchibald earlier point, what exactly would be "hidden" here if most URLs are already observable through computed styles? Jake called out fonts earlier, but afaik, those are visible via http://dev.w3.org/csswg/css-font-loading/.

**Comment by annevk** [comment]
Those observable through computed styles are only exposed if you know where to look, which can be expensive. Very different characteristics from getting a list. And e.g. `@import` is not affected by that. And it might very well be that CSS font loading has a bug too if it's claiming to be available for no-cors CSS.

**Comment by annevk** [comment]
@tabatkins ^^

**Comment by tabatkins** [comment]
FontFace objects don't expose the url (or ArrayBuffer) they were loaded from.  The FontFace itself is there as long as CSS has rights to it (font loading is CORS-restricted per spec), but you can't read anything from it.

**Comment by annevk** [comment]
@tabatkins no-cors CSS is applied to a page, but you can't access it through CSSOM (you can access computed styles). So the question is what the font API does for such a CSS resource (I suspect it shouldn't work, but that may not be defined).

**Comment by davidben** [comment]
Oh yuck. Yeah, I think I agree with Anne that we should remove these requests from SW and Resource Timing unless you add the `crossorigin` attribute. These kinds of "the contents are secret, but if they happen to parse as foo, you can execute it" security policies are super-hairy. We shouldn't add new ones.

In fact, cross-origin CSS has already bitten us in the past because the CSS parser is extremely error-tolerant. See https://www.linshunghuang.com/papers/css.pdf

**Comment by davidben** [comment]
Oh, and I believe this is the corresponding WebKit bug, since I don't see it linked in the paper:
https://bugs.webkit.org/show_bug.cgi?id=29820

**Comment by tabatkins** [comment]
Yeah, makes sense.  I'll send a notification to www-style and make the change in Font Loading.

**Comment by tabatkins** [comment]
Actually, I ran into some questions about how to implement.  I posted https://lists.w3.org/Archives/Public/www-style/2015Jul/0150.html with two possibilities for handling this.

The proposal I prefer still exposes the *existence* of FontFace objects from the tainted stylesheet, and their loading status and promise.  Everything else is hidden, tho.

(If you know the name of the font, its loading status is *already* exposed to the page via layout/timing channels. This approach would further expose the *number* of fonts present in tainted sheets and the loading status/promises of the mystery fonts, which wasn't previously available, but any further restrictions seem kinda insane to spec/implement.)

**Comment by nattokirai** [comment]
Creating new flavors of FontFace seems like overdesign to me. If style rules can't be exposed in the OM, I don't think FontFace objects should be exposed in the FontFaceSet. The ready() promise should include fonts from the inaccessible stylesheet.

> This is kinda similar to how I'd like to expose local fonts at some point.

Allowing local fonts to be enumerated exposes users to fingerprinting attacks.

**Comment by tabatkins** [comment]
> Creating new flavors of FontFace seems like overdesign to me. If style rules can't be exposed in the OM, I don't think FontFace objects should be exposed in the FontFaceSet. The ready() promise should include fonts from the inaccessible stylesheet.

I explained in the email why that's complicated and probably unworkable.  Please comment on www-style for further discussing of FontFace/etc.

> Allowing local fonts to be enumerated exposes users to fingerprinting attacks.

I didn't mention enumerating local fonts; my suggestion was for the *exact opposite*, actually.

And you can already enumerate local fonts fairly trivially through layout channels.

**Comment by horo-t** [comment]
https://crbug.com/532374 tracks this for Chromium.

**Comment by wanderview** [comment]
Some corner cases:

Consider:

1) a.com/index.html loads stylesheet at b.com/foo.css as no-cors
2) b.com/foo.css @imports stylesheet at a.com/bar.css
3) a.com/bar.css loads background-image a.com/snafu.jpg

Should SW and performance see data for snafu.jpg?  I think everything @imported under a tainted sheet should be hidden.

@igrigorik, note this is not really covered by the current language in the performance spec:

> For each resource fetched by the current browsing context, excluding resources fetched by cross-origin stylesheets fetched with no-cors policy, perform the following steps: 

This raises further issues like this:

1) a.com/index.html loads stylesheet at b.com/foo.css as no-cors
2) b.com/foo.css @imports stylesheet at a.com/bar.css
3) a.com/bar.css loads background-image a.com/snafu.jpg
4) a.com/index.html loads stylesheet at a.com/thepain.css
5) a.com/thepain.css @imports stylesheet at a.com/bar.css
6) a.com/bar.css loads background-image a.com/snafu.jpg

Here snafu.jpg should be hidden in step 3, but what happens to snafu.jpg in step 6?  It seems the stylesheet should not be shared from tainted import with non-tainted import.  I'm not sure if the image cache would prevent a network load in step 6, though.

**Comment by sicking** [comment]
I don't think we should prevent requests coming from a no-cors stylesheet from going through a SW.

Yes. I understand that this exposes data that isn't currently exposed. And yes, I understand that there is some security risk involved in that. But I don't think that automatically means that we must not do this.

In short, I think the question is more complex and nuanced than that.

Currently, if I host a file on a webserver and serve it with a "text/css" mimetype, *a lot* of the data in that file can be read by third parties.

A third party can link to that stylesheet, create an element which matches one of the rules in the stylesheet, and then use getComputedStyle to figure out the properties that were set by the rule, and what those properties were set to.

You can relatively easily use a dictionary to scan probable class names to extract all rules that simply match on a class. You can even scan all possible ascii-classnames shorter than N characters.

That means, that as an author, if I put a text/css resource on a webserver, I already need to count on that most of the bytes in that file is readable by third parties.

It's also not as simple as "exposing more data" == "less safe for developers".

Trying to send developers the message that "these here bytes are leaked, but these other bytes, in the @import rule, those are safe, feel free to put sensitive data in those URLs" is a significantly more complex than saying "the contents of the stylesheet can be read by third parties. Don't put sensitive data in stylesheets". Such complexity often leads to security bugs.

For example, the fact that background-image URL isn't exposed to a SW might make developers think that the URL is protected, when in fact .getComputedStyle leaks the very same information. I.e. it is very easy for developers to get a false sense of security. This is also made worse by the fact that developers can legitimately claim that some urls, like @import, are in fact safe.

Then there is of course the problem that the implementation complexity will most likely lead to a few security bugs here and there. We had a recent example of this where a developer had seen that CORS says that only certain values can be sent in the Content-Type header without triggering a preflight. So the developer built CSRF-protection by checking for a particular content type. However it turned out that different browsers apply slightly different rules to Content-Type parsing leading to the ability for an attacker to send a cross-site request in some browsers with a Content-Type that this website didn't expect.

The Content-Type rule in CORS is my fault. I think it was a bad decision and it has lead to security problems.

What we're talking about here is orders of magnitude more complex, and so I feel quite confident that it will lead to orders of magnitude more problems. Both security problems and other problems.

And yes, I am aware that we block access to the CSSOM for cross-origin stylesheets that weren't fetched with CORS. But that was added because at the time we didn't have the restriction that the stylesheet had to be served as "text/css". The attacks that we saw was due to websites pointing <link rel=stylesheet> at cross-origin HTML files, and then took advantage of the lax CSS parsing rules to extract information from the HTML.

So, all in all, I think the cost here is quite high to browser developers, as can be seen by all the edge cases that are being debated. I agree that there is some security value to developers, but I also think there is a very real security risk to developers (in addition to the cost of not being able to intercept CDN hosted stylesheet resources).

And again, I realize that more information is being exposed here. If your counter argument is simply that that means that we have to block these intercepts, then you might want to reread this comment.

I think we would be better served giving developers a real security tool here. I know that @annevk had a proposal some time back for a header which blocked cross-site reads of resources like images/stylesheets/scripts. Implementing something like that seems like a better way to protect the data debated here.

**Comment by sicking** [comment]
FWIW, what I think we should do is:

* Treat resources served as "text/css" as "server has indicated that the contents of this file contains no private information"
* Add a header like "sec-from: same-origin" which allows a website to easily indicate that a given stylesheet should not be loadable cross-site.

The second bullet has been discussed before now but has the nice property that it would actually protect the information debated here on all UAs. Old UAs that don't support the property also don't support SW.

**Comment by wanderview** [comment]
Some IRC discussion:

 http://logs.glob.uno/?c=freenode%23whatwg#c974181

**Comment by jakearchibald** [comment]
F2F: we need browser security people to weigh in on this

**Comment by yoavweiss** [comment]
Did sec folks comment on this?

**Comment by wanderview** [comment]
AFAIK, no they have not.  Its been 6 months and this API has been released for 1+ years in one form or another.  At this point I feel like we should probably just close this as WONTFIX.  Of course, if security knowledgeable people feel strongly otherwise that feedback is quite welcome.

**Comment by annevk** [comment]
I still think poking holes in SOP is a bad idea.

**Comment by sicking** [comment]
A slightly larger nice round hole is better than a irregular jagged hole
which in practice can't be depended on to be any smaller than the round
hole.

**Comment by jakearchibald** [comment]
@mikewest we need help making a call here. This has been in the wild for a long time and doesn't seem to have caused a problem, but we're worried.

We should also add some use counters for this.

@domenic & @jakearchibald need to look through https://www.linshunghuang.com/papers/css.pdf

**Comment by jakearchibald** [comment]
We should have a VC about this, with security people.

**Comment by hober** [comment]
cc @johnwilander

**Comment by johnwilander** [comment]
Small note: WebKit does not CORS-restrict font loading (https://bugs.webkit.org/show_bug.cgi?id=86817).

**Comment by tdresser** [comment]
What's the next step here?

**Comment by annevk** [comment]
I suspect it's enshrining a security hole because nobody acted.

**Comment by jakearchibald** [comment]
Action: @slightlyoff is cornering security folks RIGHT NOW.

**Comment by jakearchibald** [comment]
F2F: We need to watch out for this if workers become subresoruces, and allow cross origin with no-cors.

**Comment by shhnjk** [comment]
1. Add "css" (and "worker" if required) in [request initiator](https://fetch.spec.whatwg.org/#concept-request-initiator).
2. Add opaque-subresource request (near [subresource request](https://fetch.spec.whatwg.org/#subresource-request)).
>A opaque-subresource request is a request whose initiator was fetched with [request mode](https://fetch.spec.whatwg.org/#concept-request-mode) "no-cors" and response was cross-origin (should be replaced with better sentence).
3. Add following step between step 10 and step 11 of [handle fetch](https://w3c.github.io/ServiceWorker/#handle-fetch)
> If request is a opaque-subresource request, then: Return null.

This should do. Let's fix this.

**Comment by wanderview** [comment]
> response was cross-origin

@shhnjk, what is this cross-origin compared to?  The client initiating the request?  Or the document/stylesheet initiating the request?

I'm just curious if we care about the case where we have:

1. Document with origin foo.com load stylesheet "foo.com/A.css"
2. Stylesheet "foo.com/A.css" does `@import("bar.com/B.css")`
2. Stylesheet "bar.com/B.css" does `@import("foo.com/C.css")`

Is the "foo.com/C.css" load considered same-origin or cross-origin for the purposes of this check?  If its considered same-origin then it seems some information about the cross-origin "bar.com/B.css" has been leaked.

The situation is similar to how we handle CORS and redirects.  Once you redirect through a cross-origin that request is tainted as cross-origin even if its redirects back to same-origin.

**Comment by shhnjk** [comment]
>what is this cross-origin compared to? The client initiating the request? Or the document/stylesheet initiating the request?

Client. We should not check same-origin/cross-origin based on stylesheet.

>I'm just curious if we care about the case where we have:
1 Document with origin foo.com load stylesheet "foo.com/A.css"
2 Stylesheet "foo.com/A.css" does @import("bar.com/B.css")
3 Stylesheet "bar.com/B.css" does @import("foo.com/C.css")
Is the "foo.com/C.css" load considered same-origin or cross-origin for the purposes of this check? If its considered same-origin then it seems some information about the cross-origin "bar.com/B.css" has been leaked.

Step 3:
Initiaor: "bar.com/B.css"
Initiaor was fetched with "no-cors": true
Initiator was cross-origin: true
return null

So this will not leak the info. But SW point of view, you might want to serve it from cache though. My solution is based on security not performance.

BTW, `Add "css" (and "worker" if required) in request initiator` might not be required. we just need to check the intiator's destination was "style" (or "worker").

**Comment by yoavweiss** [comment]
Any news on this? This is blocking https://github.com/w3c/resource-timing/issues/70

**Comment by rniwa** [comment]
WebKit is wiling to change its behavior to avoid exposing resources fetched by non-CORS cross-origin CSS resources. We don't think it's okay to leak information about non-CORS cross-origin resources like this.

**Comment by jakearchibald** [comment]
My concerns about this:

It's difficult for developers to understand which parts of their CSS is protected when loaded cross-origin:

**//cross-origin/user.css**

```css
@import 'user/styles/david-smith.css';
```

**//cross-origin/user/styles/david-smith.css**

```css
@font-face {
  font-family: 'david-loves-consolas';
  src: url('/fonts/consolas-v123.woff2');
}

body {
  font-family: david-loves-consolas;
}

.user-avatar {
  background: url('/images/me-and-samantha.jpg');
}

.davids-mother-is-called-jane {
  background: green;
}
```

Assuming we make the change proposed in this issue, what's private?

We can look at the computed style for body, and determine that the user is called david, and they love consolas. Again, using computed styles, we can determine they have a friend or relative called samantha, who is displayed in the image. We could use guesswork to figure out David's mother is called Jane.

The bits of information that this PR protects is David's surname, as it's only in the `@import` which isn't usually exposed. We also protect the version of the font (I think? Or is this exposed elsewhere?).

My advice to developers would be "Don't put private data in CSS". I realise we do our best to hide the source, as we do with JS, but we don't hide fetches from JS.

Also, as well as this change breaking existing sites, it may not be possible to fix. Although you can load a stylesheet with CORS, there's no way to mark additional imports as CORS.

**Comment by yoavweiss** [comment]
FWIW, https://jsbin.com/pigihubuxa/edit?html,output show that at least Chrome and Safari do expose CSSOM (and the background image URL info) for default cross-origin CSS fetches. Firefox does not.

**Comment by jakearchibald** [comment]
As an aside, isn't script and CSS source highly vulnerable to meltdown/spectre?

**Comment by davidben** [comment]
Indeed! See also CORB and friends. The legacy thing where the web platform has all these "you can execute this but not read through contents" CORS bypasses are kind of a disaster. I need to swap this particular issue back in, but having to sacrifice ergonomics here to plug a vulnerability would not be at all surprising to me.

**Comment by annevk** [comment]
SOP bypasses* (this seems an increasingly futile battle, but I'll keep trying)

**Comment by mattto** [comment]
F2F: This was discussed partly in WebAppSec and in a dedicated breakout session:
- https://www.w3.org/2018/10/23-webappsec-minutes.html#item12
- https://www.w3.org/2018/10/24-csssop-minutes.html

If I may attempt a summary.

Arguments in favor of making the requests skip the service worker:
- It's arguably a SOP bypass.
- There have been real instances of CSS leaking private data, ca. 2011 (Eric Lawrence to try to provide more details about this).
- Firefox changed to at least partially block this for Resource Timing: https://bugzilla.mozilla.org/show_bug.cgi?id=1180145
- Allowing a SOP bypass because it's already exposed via other APIs or side-effects seems unsatisfying; not moving where we want to be.

Arguments in favor of going to the service worker:
- The information is already exposed via Resource Timing; getComputedStyle(); and side-effects of the import itself on geometry/etc.
- Changing the behavior now could break sites.
- It would become difficult to explain to developers when subresource requests go to the service worker and when they don't.
- There would be some complexity in specing/implementing skipping, due to nested imports and redirects.

Observations:
- Any new APIs won't make subresource requests from no-cors contexts. If we were to start from scratch, we'd have made CSS cors-only.
- A core problem is there is no way for developers to make cors `import` requests. You can only do cors from `<link>`.

Outcomes:
- We want to timebox this decision (there were some references to mid-November, or this year).
- Youenn (WebKit) is to check internally to try to make a decision about what they want to do.
- @wanderview (Chrome) volunteered to add metrics to estimate how many sites would be affected.

(I don't know how to cc Eric L or Youenn on GitHub.)

**Comment by yoavweiss** [comment]
cc @youennf @ericlaw1979

**Comment by youennf** [comment]
> * A core problem is there is no way for developers to make cors `import` requests. You can only do cors from `<link>`.

Is there a corresponding GitHub issue tracking this?

**Comment by aliams** [comment]
@wanderview were you able to get metrics for this yet by any chance?

**Comment by wanderview** [comment]
I spent a couple days working on this,  but it turned out to be more complex than I had time for.  The issue I ran into is that chrome does not initiate the loads for things like background images, etc, until they are used by a style flush.  This is quite a bit after the sheet was loaded and parsed, so none of the tainting information is available.  That information would have to be plumbed though.  It was a bit too much for me since I'm not familiar with chrome's style system.

**Comment by mattto** [comment]
I've just landed some metrics at https://crbug.com/898497. Very early data from the past few Chrome Canaries suggests that 0.87% of page loads involved a service worker intercepting a request from a no-cors stylesheet. 14% of page loads involved a controller service worker at all, and of those page loads 6% had a no-cors stylesheet making a request.

**Comment by mattto** [comment]
My personal interpretation of this is that it's likely in the range where we can deprecate, and the implementation complexity doesn't seem so bad (Blink already tracked the "origin-clean" bit for nested imports and redirects, AFAICT). Therefore, of the arguments against deprecating at https://github.com/w3c/ServiceWorker/issues/719#issuecomment-433819735, I think the strongest one is developer confusion. Not sure how much of a blocker that should be.

**Comment by youennf** [comment]
FWIW, Safari Tech Preview and Safari latest betas skip service worker for those loads.
So far, we have not heard of breakage.

**Comment by annevk** [comment]
That is exciting, I've reopened https://bugzilla.mozilla.org/show_bug.cgi?id=1201160 so we can reconsider this for Firefox.

**Comment by youennf** [comment]
We received no complaint so far so this seems web compatible.
It could be interesting to see how affected the offline experience would be for those sites.

**Comment by mfalken** [comment]
Good idea. The [chromestatus page](https://chromestatus.com/metrics/feature/timeline/popularity/2941) lists some sample URLs that are affected.

Would it be possible to see if any of those have a degraded offline experience on a version of Safari that has the change?

**Comment by asutherland** [comment]
From a spec perspective, the changes for this would actually happen elsewhere and result in resources being fetched with a [service-workers mode](https://fetch.spec.whatwg.org/#request-service-workers-mode) of "none", yes?

**Comment by hiring-committee-github-viewer** [comment]
Agreed, it would be part of CSS I imagine.

**Comment by mfalken** [comment]
(fixing account) Agreed, it would be part of CSS I imagine.

**Comment by annevk** [comment]
Yeah, though this repository should probably take ownership of getting it tested as CSS hasn't even integrated with Fetch and its integration with HTML is also patchy.

**Comment by mfalken** [comment]
(belated) Notes from [Service Worker WG meeting on Nov 20](https://docs.google.com/document/d/1ybS1q2HCPh3bNNOkjGpAPFug19A2BsIxYEi-i6lrB1w/edit?usp=sharing):

* Safari Tech Preview is skipping service workers for no-cors requests from CSS.
* We should see whether the offline experience of sites here https://chromestatus.com/metrics/feature/timeline/popularity/2941 are impacted in Safari Tech Preview.
* Affecting 1.5% of page loads will be considered very high by Blink API Owners.
* CSS would need to integrate with Fetch to write the spec for this.

**Comment by sicking** [comment]
I'm curious how we are thinking about the issue here in relation to the fact that a lot of the same information is exposed through the CSSOM in most major browsers, Safari and Chrome included [1].

What would be the security win if we just fix this issue without also fixing the CSSOM issue?

Are there plans to also fix the CSSOM issue? Has any experiments been done to see what percentage of websites would be impacted? Or any experiments shipped with that support disabled?

I'm also curious if anyone has heard of this issue leading to security problems? This design has been shipping for several years now so we've had opportunity to see if this is a real-world problem or just a theoretical one.

I think it's well established that no-cors cross-site stylesheets has lead to security issues (as referenced in [2]). However the question at hand in this issue is if locking down URL access while still exposing a lot of other properties provides meaningful security benefits. 

[1] https://jsbin.com/pigihubuxa/edit?html,output
[2] https://github.com/w3c/ServiceWorker/issues/719#issuecomment-433819735

**Comment by yoshisatoyanagisawa** [comment]
I revisited https://chromestatus.com/metrics/feature/timeline/popularity/2941 told in https://github.com/w3c/ServiceWorker/issues/719#issuecomment-765175844.
The latest data says approx. 3.5%.

**Comment by youennf** [comment]
I mentionned during the discussion that WebKit should be allowed to skip SW there.
I guess this would be accomplished by specs calling fetch to allow setting service worker mode to `none` in those specific cases.

**Comment by monica-ch** [comment]
Discussed in [TPAC 2025](https://docs.google.com/document/d/1fOjXiGUFf_krkfYaWf5drbqzY7CTxn7EcDYr5ZWjzFg/edit?tab=t.0)

Safari already hides no-cors CSS subresource fetches from service workers, and it would be helpful for the Fetch spec to explicitly allow this with a permissive “may,” so engines can adopt similar behavior if needed. Longer-term, CSS’s CORS support gives authors a path to opt in when they want these requests visible to service workers.

**Comment by annevk** [comment]
This isn't really up to Fetch I think. CSS needs to define how it integrates with Fetch and avoid this SOP hole in the way Youenn suggests.
